bin/clang/clangGlobals.h

```diff
@@ -50,10 +57,116 @@ namespace clang {
   class Sema;
 }
 
-namespace sst {
-  extern clang::CompilerInstance *activeCompiler;
-  extern clang::ASTContext *activeASTContext;
-  extern clang::SourceManager *activeSourceManger;
-  extern clang::LangOptions *activeLangOpts;
-  extern clang::Sema *activeSema;
+namespace modes {
+
+enum ModeMask {
+  ENCAPSULATE = 1 << 0,
+  MEMOIZE = 1 << 1,
+  SKELETONIZE = 1 << 2,
+  SHADOWIZE = 1 << 3,
+  PUPPETIZE = 1 << 4
+};
+
+enum Mode {
+  ENCAPSULATE_MODE = 0,
+  MEMOIZE_MODE = 1,
+  SKELETONIZE_MODE = 2,
+  SHADOWIZE_MODE = 3,
+  PUPPETIZE_MODE = 4,
+  NUM_MODES
+};
+
 }
+
+class SkeletonASTVisitor;
+class FirstPassASTVisitor;
+struct SSTPragma;
+class SSTReplacePragma;
+class SSTNullVariablePragma;
+struct PragmaConfig {
+  std::map<clang::Decl*,SSTNullVariablePragma*> nullVariables;
+  std::map<clang::FunctionDecl*,std::set<SSTPragma*>> functionPragmas;
+  std::map<const clang::DeclContext*,SSTNullVariablePragma*> nullSafeFunctions;
```

Comment:
Should we switch these to the llvm optimized structures?

Reply:
Not yet. Maybe after this PR.